java/arcs/android/host/ArcHostHelper.kt

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.android.host
+
+import android.app.Service
+import android.content.ComponentName
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import android.os.Parcelable
+import android.os.ResultReceiver
+import arcs.core.host.ArcHost
+import arcs.core.host.ParticleIdentifier
+import arcs.core.host.PlanPartition
+import java.io.Serializable
+import kotlin.reflect.KClass
+import kotlinx.coroutines.CoroutineName
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+
+/**
+ * Tool which can be used by [ArcHost]s to handle [Intent] based API calls, as well as
+ * send API calls.
+ *
+ * ## Example Usage:
+ *
+ * ```kotlin
+ * class MyService : Service() {
+ *     private val myHelper: ArcHostHelper by lazy {
+ *         ArcHostHelper(MyArcHost())
+ *     }
+ *
+ *     class MyArcHost : AbstractArcHost() {
+ *       override suspend fun onStartArc(plan: PlanPartition) {
+ *         // ...
+ *       }
+ *     }
+ *
+ *     override fun onStartCommand(intent: Intent?, flags: Int, startId: Int) {
+ *         val result = super.onStartCommand(intent, flags, startId)
+ *         myHelper.onStartCommand(intent)
+ *         return result
+ *     }
+ * }
+ * ```
+ */
+class ArcHostHelper(private val service: Service, private val arcHost: ArcHost) {
+
+    private val job = Job() + Dispatchers.IO + CoroutineName("ArcHostHelper")
+
+    /**
+     * Determines whether or not the given [Intent] represents a call to an [ArcHost] and invokes
+     * the appropriate interface methods.
+     */
+    fun onStartCommand(intent: Intent?) = CoroutineScope(job).launch {
+        onStartCommandSuspendable(intent)
+    }
+
+    // VisibleForTesting
+    suspend fun onStartCommandSuspendable(intent: Intent?) {
+        // Ignore other actions
+        if (intent?.action?.startsWith(ARC_HOST_INTENT) != true) return
+
+        // Ignore Intent when it doesn't target our Service
+        if (intent.component?.equals(ComponentName(service, service::class.java)) != true) return
+
+        when (intent.getSerializableExtra(OPERATION) as Operation?) {
+            Operation.START_ARC -> runWithResult(intent) {
+                arcHost.startArc(
+                    intent.getParcelableExtra<ParcelablePlanPartition?>(
+                        OPERATION_ARG
+                    )?.actual!!
```

Comment:
This (and the flavors below) will throw a NPE if the extra is not available. It might be better to return silently in that case.

My reasoning for returning silently is to make it so a malicious actor can't DOS a service by forcing it to crash over and over by passing an invalid ArcHost command intent.

Reply:
I think I'll add an OPERATION_ERROR to the Bundle and allow a caller to trap it, so that the Allocator can know when startArc calls fail.

Reply:
I'll handle this in a follow up. I made it handle it silently for now.